### PR TITLE
Smarter item load ordering during ProduceRestoreCollections

### PR DESCRIPTION
**Pull request by ashmrtn**
Optimize loading items for restore a little bit by first grouping items
by directory, loading the directory once, and then loading all items
from the loaded directory. This brings item loading on my local machine
(communicating with remote S3) down to ~1.5min/1k items

Future improvements could lazily load items as they're returned in the
Items() call of each collection but doing so would change the semantics
of ProduceRestoreCollections() (specifically item not found errors would
be returned during Items() instead of during
ProduceRestoreCollections())

The kopia data collection has also been updated to hold onto a
reference to the folder it corresponds to. This folder reference is
used to service Fetch() calls

---

#### Does this PR need a docs update or release note?

- [x] :white_check_mark: Yes, it's included
- [ ] :clock1: Yes, but in a later PR
- [ ] :no_entry: No

#### Type of change

- [x] :sunflower: Feature
- [ ] :bug: Bugfix
- [ ] :world_map: Documentation
- [ ] :robot: Supportability/Tests
- [ ] :computer: CI/Deployment
- [ ] :broom: Tech Debt/Cleanup

#### Issue(s)

* #3293

#### Test Plan

- [ ] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/kopia/data_collection.go
Comment (ryanfkeepers) [inline src/internal/kopia/data_collection.go:74]:
```suggestion
		return nil, clues.Wrap(errNoRestorePath, "unknown item").WithClues(ctx)
```

**Review by ryanfkeepers — COMMENTED** [review]
Generally LGTM.  I can approve if wanted, but I also think it's a good idea for @pandeyabs to review as well before merging.  For his opportunity to dig into/keep up with kopia changes, if nothing else.

**Review by pandeyabs — COMMENTED** [review]
@ashmrtn I'll take a look at this PR later today. Since I am new to kopia, could you please tell me how were we loading items before this PR? I can go through the code but this would be easier. Trying to understand how this PR is improving item load time!

**Comment by ashmrtn** [comment]
> I'll take a look at this PR later today.

sounds good, thanks!

>Since I am new to kopia, could you please tell me how were we loading items before this PR? I can go through the code but this would be easier. Trying to understand how this PR is improving item load time!

Previously the restore logic was pretty naive. The `ProduceRestoreCollections` function was passed a set of `path.Path` entries, each of which denotes the precise location in a kopia snapshot of the item to load (this info is sourced from RepoRef in backup details). The old logic would take each of those paths and [ask](https://github.com/alcionai/corso/blob/d067a7948326a42ade0b5d1d6f39d062f9dc5b78/src/internal/kopia/wrapper.go#L344) kopia to [fetch the nested entry](https://pkg.go.dev/github.com/kopia/kopia/snapshot/snapshotfs#GetNestedEntry), starting at the snapshot root. This would cause lots of traversals of the snapshot's hierarchy and apparently kopia doesn't cache those intermediate results (well or at all, not sure which at the moment)

This new PR adds a different call to [Directory.Child](https://pkg.go.dev/github.com/kopia/kopia/fs#Directory) to get entries directly from a loaded directory (the directory is still loaded by asking for a nested entry from the snapshot root). That alone cuts down on hierarchy traversals, thus helping load times

File: src/internal/kopia/wrapper.go
Comment (pandeyabs) [inline src/internal/kopia/wrapper.go:372]: Could you add a short description for this function as well?

File: src/internal/kopia/data_collection.go
Comment (pandeyabs) [inline src/internal/kopia/data_collection.go:91]: Does this need a `size` < 0 error check?

**Review by pandeyabs — APPROVED** [review]
Added some minor follow ups, but overall looks good to me!

File: src/internal/kopia/data_collection.go
Comment (ashmrtn) [reply src/internal/kopia/data_collection.go:91]: hmmm probably a good check to have. I'll add a log message so we know it happened. Ideally it shouldn't though because this should be the version that's injected by the backup logic

**Pull request merged**
